src/SAML2/Attribute/AttributeSet.php

```diff
@@ -51,6 +69,15 @@ private function __construct()
     }
 
     /**
+     * @param AttributeSetFilterInterface $attributeFilter
+     * @return AttributeSet
+     */
+    public function apply(AttributeSetFilterInterface $attributeFilter)
+    {
+        return self::create($attributeFilter->applyOn($this->attributes));
```

Comment:
But why don't we just use some kind of filter predicate, rather than an external filter? It could be as simple as `self::create(array_filter($this->attributes, [$filter, 'accepts']))`. This way we don't pass an array of Attributes while we have a perfectly typed AttributeSet. Or is knowledge of multiple attributes required for the ARP at a later point? Even then, working with an AttributeSet seems nicer than a plain array.